net/pfSense-pkg-WireGuard/files/usr/local/pkg/wireguard/wg_guiconfig.inc

```diff
@@ -0,0 +1,338 @@
+	
+	// If end time is not supplied, use the current time instead
+	$end_time = is_null($end_time) ? time() : $end_time;
+
+	// If start time is zero, proceed unless we are actually interested in calculating time from epoch, probably not...
+	$start_time = ($start_time > 0) ? $start_time : ($from_epoch ? $start_time : $end_time);
+
+	$diff_time = abs($end_time - $start_time);
+
+	$tokens = array (
+		31536000 => 'year',
+		2592000 => 'month',
+		604800 => 'week',
+		86400 => 'day',
+		3600 => 'hour',
+		60 => 'minute',
+		1 => 'second'
+	);
+
+	foreach ($tokens as $unit => $text) {
+
+		if ($diff_time < $unit) {
+			
+			continue;
+			
+		}
+
+		$num_of_units = floor($diff_time / $unit);
+
+		$plural = ($num_of_units > 1) ? 's' : '';
+		
+		return "{$num_of_units} {$text}{$plural}";
+
+	}
+
+	return "never";
+
+}
+
+// Returns the appropriate text input field type for secrets
+function wg_secret_input_type() {
+	global $wgg;
+
+	wg_globals();
+
+	if (isset($wgg['config']['hide_secrets']) && $wgg['config']['hide_secrets'] =='yes') {
+
+		return 'password';
+
+	}
+
+	return 'text';
+	
+}
+
+// Gets the appropriate class based on whether or not the target is enabled or disabled
+function wg_entrystatus_class($target_device) {
+	global $wgg;
+
+	if (isset($target_device) && 
+		$target_device['enabled'] == 'yes' &&
+			is_module_loaded($wgg['kmod'])) {
+
+		return 'enabled';
+
+	}
+
+	return 'disabled';
+
+}
+
+// Generates a toggle icon link based on the status of the target
+function wg_generate_toggle_icon_link($target_device, $title = '', $href = '#', $usepost = true, $icon_enabled = 'fa-ban', $icon_disabled = 'fa-check-square-o') {
+
+	if (isset($target_device)) {
+	
+		$s_icon = (isset($target_device['enabled']) && $target_device['enabled'] == 'yes') ? $icon_enabled : $icon_disabled;
+
+		$s_title = gettext($title);
+
+		$s_usepost = $usepost ? 'usepost' : '';
+
+		return "<a class=\"fa {$s_icon}\" title=\"{$s_title}\" href=\"{$href}\" {$s_usepost}></a>";
+
+	}
+
+	return null;
+
+}
+
+function wg_generate_tunnel_address_popup_link($tunnel_name) {
+	global $wgg;
+
+	$tun_idx = wg_get_tunnel_array_index($tunnel_name);
+
+	if (isset($wgg['tunnels'][$tun_idx]) && is_array($wgg['tunnels'][$tun_idx])) {
+
+		$tunnel = $wgg['tunnels'][$tun_idx];
+
+		$addresses = $tunnel['addresses']['row'];
+
+		if (!is_wg_tunnel_assigned($tunnel['name'])) {
+			
+			if (is_array($addresses) && count($addresses) > 0) {
+
+				$href = "vpn_wg_tunnels_edit.php?tun={$tunnel['name']}";
+
+				$extras = count($addresses) - 1;
+
+				if ($extras > 0) {
+
+					$data_html = "<table><thead><th>Address</th><th style='padding-left: 10px;'>Description</th></thead><tbody>";
+		
+					foreach ($addresses as $address) {
+		
+						$data_html .= "<tr><td>{$address['address']}/{$address['mask']}</td><td style='padding-left: 10px;'>{$address['descr']}</td></tr>";
+		
+					}
+		
+					$data_html .= '</tbody></table>';
+		
+					$hint = "{$addresses[0]['address']} (+{$extras})";
```

Comment:
The contents of this string should be encoded with `htmlspecialchars()` before printing, either here or in the line below.